### PR TITLE
add dataset heartbeat

### PR DESCRIPTION
**Pull request by NakulManchanda**
This PR adds:
- POST datasetHeartbeat
- addZssBackgroundTask function
- reading new configuration from environment - ZSS_DATASET_HEARTBEAT, ZSS_DATASET_EXPIRY
- new file - datasetlock.h

We are using new STC BASE Module for monitoring, and created interface to other plugins and services can use it in future too.

### CODE REVIEW

File: c/zss.c [hunk 1]
```diff
@@ -389,6 +394,25 @@ static JsonObject *readServerSettings(ShortLivedHeap *slh, const char *filename)
...
+  stcRegisterModule(
+    base,
+    STC_MODULE_ZSS,
+    server,
+    NULL,
+    NULL,
+    NULL,
+    processZssBackgroundHandler
+  );
+
```
Comment (NakulManchanda) [inline c/zss.c:410]: @ifakhrutdinov - here is our registration to STC Module
we call `stcRegisterModule`, with new module constant `STC_MODULE_ZSS` to process background handler `processZssBackgroundHandler`

File: c/datasetHeartbeatService.c [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+#include "zssLogging.h"
+
+#include "datasetService.h"
+#include "datasetHeartbeatService.h"
+
+#ifdef __ZOWE_OS_ZOS
+
+static int serveDatasetHeartbeat(HttpService *service, HttpResponse *response){
+  time_t T;
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "begin %s\n", __FUNCTION__);
```
Comment (jordanfilteau1995) [inline c/datasetHeartbeatService.c:51]: I doubt we want 'INFO' for these. Please don't forget to revert them if you set them for development.

File: c/datasetHeartbeatService.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+static int serveDatasetHeartbeat(HttpService *service, HttpResponse *response){
+  time_t T;
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "begin %s\n", __FUNCTION__); 
+  HttpRequest *request = response->request;
+  if (!strcmp(request->method, methodPOST)) { 
+    resetTimeInHbt(request->username);
+
+    zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "user name=%s\n", request->username);
+    time(&T);
```
Comment (jordanfilteau1995) [inline c/datasetHeartbeatService.c:57]: I don't think this is available in METAL. In that case, the #define are kind of useless since it won't compile anyways. Does https://github.com/zowe/zowe-common-c/blob/master/c/timeutls.c do what you want?

Comment (jordanfilteau1995) [reply c/datasetHeartbeatService.c:57]: I didn't test it, but I usually go off this list:

https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.2.0/com.ibm.zos.v2r2.ccrug00/fnct.htm

File: c/zss.c [hunk 4]
```diff
@@ -94,6 +96,8 @@ char productVersion[40];
 
 static JsonObject *MVD_SETTINGS = NULL;
 static int traceLevel = 0;
+extern int heartbeat_loop_time;
```
Comment (jordanfilteau1995) [inline c/zss.c:99]: This is okay, but we usually use camelCase for variable names everywhere else, so it might be best to follow the style already in the file.

File: c/datasetHeartbeatService.c [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    jsonPrinter *out = respondWithJsonPrinter(response);
+    setContentType(response, "text/json");
+    setResponseStatus(response, 405, "Method Not Allowed");
+    addStringHeader(response, "Server", "jdmfws");
+    addStringHeader(response, "Transfer-Encoding", "chunked");
+    addStringHeader(response, "Allow", "POST");
+    writeHeader(response);
+
+    jsonStart(out);
+    jsonEnd(out);
```
Comment (ifakhrutdinov) [inline c/datasetHeartbeatService.c:75]: Can you reuse some of the functions in `zowe-common-c/c/httpserver.c`? Please have a look at `respondWithError()`, `respondWithMessage()` and `setDefaultJSONRESTHeaders()`.

File: c/datasetHeartbeatService.c [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    finishResponse(response);
+  }
+
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "end %s\n", __FUNCTION__);
```
Comment (ifakhrutdinov) [inline c/datasetHeartbeatService.c:80]: What is the point in having 2 almost identical messages?

File: c/datasetHeartbeatService.c [hunk 7]
```diff
@@ -0,0 +1,110 @@
+
```
Comment (ifakhrutdinov) [inline c/datasetHeartbeatService.c:1]: Do you really need a separate source file and installer for this service? Have a look at `c/dataService.c`. Perhaps you can extend that.

File: c/datasetService.c [hunk 8]
```diff
@@ -86,9 +86,8 @@ static int serveDatasetMetadata(HttpService *service, HttpResponse *response) {
 }
 
 static int serveDatasetContents(HttpService *service, HttpResponse *response){
-  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_DEBUG2, "begin %s\n", __FUNCTION__);
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "begin %s\n", __FUNCTION__);
```
Comment (ifakhrutdinov) [inline c/datasetService.c:89]: What's the purpose of this change?

File: c/datasetHeartbeatService.c [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+void installDatasetHeartbeatService(HttpServer *server) {
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, ZSS_LOG_INSTALL_MSG, "dataset Heartbeat");
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "begin %s\n", __FUNCTION__);  
+
+  HttpService *httpService = makeGeneratedService("datasetHeartbeart", "/datasetHeartbeat/**");
+  httpService->authType = SERVICE_AUTH_NATIVE_WITH_SESSION_TOKEN;
+  httpService->runInSubtask = TRUE;  
+  httpService->doImpersonation = TRUE;
```
Comment (ifakhrutdinov) [inline c/datasetHeartbeatService.c:93]: What is impersonation used in this service for?

File: c/datasetHeartbeatService.c [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "end %s\n", __FUNCTION__);  
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, "Returning from %s\n", __FUNCTION__); 
+  return 0;
+}
+
+
+void installDatasetHeartbeatService(HttpServer *server) {
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, ZSS_LOG_INSTALL_MSG, "dataset Heartbeat");
```
Comment (ifakhrutdinov) [inline c/datasetHeartbeatService.c:87]: One of these is probably extraneous. Neither of these should be `ZOWE_LOG_INFO`.

File: c/zss.c [hunk 11]
```diff
@@ -389,6 +394,25 @@ static JsonObject *readServerSettings(ShortLivedHeap *slh, const char *filename)
...
+    STC_MODULE_ZSS,
+    server,
+    NULL,
+    NULL,
+    NULL,
+    processZssBackgroundHandler
+  );
+
+  for(int i=0; i < N_TASK_TABLE_ENTRIES; i++) {
+    task_list[i].id = 0;  /* initialise */
```
Comment (ifakhrutdinov) [inline c/zss.c:412]: Where does this live? Is this global somewhere?

**Review by ifakhrutdinov — CHANGES_REQUESTED** [review]
Please see my comments. Once I've gone through the zowe-common-c PR I may have more questions here.

File: c/zss.c [hunk 12]
```diff
@@ -1226,6 +1266,8 @@ int main(int argc, char **argv){
       installDatasetMetadataService(server);
       installDatasetContentsService(server);
       installDatasetEnqueueService(server);
+      installDatasetHeartbeatService(server);
+      initDatasetLocking(server);
```
Comment (ifakhrutdinov) [inline c/zss.c:1270]: This doesn't belong in zss.c, it's specific to the dataset services. All ZSS needs to care about here is the install functions.

File: c/datasetHeartbeatService.c (hunk 2 above)
Comment (NakulManchanda) [reply c/datasetHeartbeatService.c]: made it DEBUG - actually remove file just added function in already existing datasetService.c

File: c/datasetHeartbeatService.c (hunk 3 above)
Comment (NakulManchanda) [reply c/datasetHeartbeatService.c]: seems to work in my z/OS as expected. I can give a demo if required

File: c/zss.c [hunk 13]
```diff
@@ -1173,7 +1212,7 @@ int main(int argc, char **argv){
   ShortLivedHeap *slh = makeShortLivedHeap(0x40000, 0x40);
   JsonObject *envSettings = readEnvSettings("ZWED");
   JsonObject *mvdSettings = readServerSettings(slh, serverConfigFile);
-
+  readDatasetSettings();
```
[the remark this answers is not in the document]
Comment (NakulManchanda) [reply c/zss.c]: yes, moved inside datasetService.c, thanks

File: c/zss.c [hunk 14]
```diff
@@ -816,6 +840,21 @@ static void printZISStatus(HttpServer *server) {
 
 }
 
+static void readDatasetSettings() {
+  JsonObject *datasetLockSettings = readEnvSettings("ZSS_DATASET");
```
[the remark this answers is not in the document]
Comment (NakulManchanda) [reply c/zss.c]: added NULL check

File: c/datasetHeartbeatService.c [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+#include "logging.h"
+#include "zssLogging.h"
+
+#include "datasetService.h"
+#include "datasetHeartbeatService.h"
+
+#ifdef __ZOWE_OS_ZOS
+
+static int serveDatasetHeartbeat(HttpService *service, HttpResponse *response){
+  time_t T;
```
[the remark this answers is not in the document]
Comment (NakulManchanda) [reply c/datasetHeartbeatService.c]: renamed it to curr_time

File: c/datasetHeartbeatService.c (hunk 5 above)
Comment (NakulManchanda) [reply c/datasetHeartbeatService.c]: using respondWith functions now

File: c/datasetHeartbeatService.c (hunk 6 above)
Comment (NakulManchanda) [reply c/datasetHeartbeatService.c]: removed it i believe

File: c/zss.c [hunk 16]
```diff
@@ -816,6 +840,21 @@ static void printZISStatus(HttpServer *server) {
...
+  int heartbeat = jsonObjectGetNumber(datasetLockSettings, "ZSS_DATASET_HEARTBEAT");
+  if(heartbeat>0) {
```
[the remark this answers is not in the document]
Comment (NakulManchanda) [reply c/zss.c]: changed all if's and for's, should be better now

File: c/datasetHeartbeatService.c (hunk 9 above)
Comment (NakulManchanda) [reply c/datasetHeartbeatService.c]: made it false

File: c/datasetHeartbeatService.c (hunk 10 above)
Comment (NakulManchanda) [reply c/datasetHeartbeatService.c]: removed

File: c/zss.c (hunk 11 above)
Comment (NakulManchanda) [reply c/zss.c:410]: yes this is global right now

File: c/zss.c (hunk 12 above)
Comment (NakulManchanda) [reply c/zss.c]: moved inside datsetService.c
